### PR TITLE
Fix PipelineRun premature failure when TaskRun recovers after pod eviction

### LINKED ISSUE #9467: PipelineRun permanently fails when TaskRun recovers after pod eviction and retriggers
# Expected Behavior
When a TaskRun's pod is evicted due to node resource pressure (exit code 137) and the controller successfully creates a replacement pod that eventually succeeds, the parent PipelineRun should reflect the TaskRun's final successful status.

# Actual Behavior
We observed unusual behavior with a TaskRun and PipelineRun. The TaskRun entered an Unknown state with reason "ExceededNodeResources" and message "TaskRun Pod exceeded available resources", where it remained stuck for approximately 5 minutes waiting for cluster resources.
The TaskRun then transitioned to Pending state with "PodReadyToStartContainers: False". Just one second later, the init container failed with exit code 137 and message "init container failed, 'prepare' exited with code 137: ContainerStatusUnknown".
This init container failure caused the PipelineRun to be marked as Failed with status "Tasks Completed: 4 (Failed: 1, Cancelled 0), Skipped: 14". All controllers (results, chains, kubearchive) started reconciling this failure.
However, the weird thing is that immediately after this, the TaskRun retried or something happened, and it went back to Unknown state with
"ExceededNodeResources" again. The TaskRun then entered Running state and eventually succeeded with "All Steps have completed executing".
The issue is that the PipelineRun remains in Failed state even though the TaskRun eventually succeeded.

# Additional Info
Observed Timeline
TaskRun controller logs:
```
  10:01:07  TaskRun → Unknown/ExceededNodeResources
            "TaskRun Pod exceeded available resources"

  10:01:18  TaskRun → Unknown/Pending
            "pod status \"PodReadyToStartContainers\":\"False\"; message: \"\""

  10:01:19  TaskRun → False/Failed
            "init container failed, \"prepare\" exited with code 137: ContainerStatusUnknown"

  10:01:20  TaskRun → Unknown/ExceededNodeResources   ← TaskRun retriggers with new pod
            "TaskRun Pod exceeded available resources"

  10:01:25  TaskRun → Unknown/Pending
            "Pending"

  10:01:33  TaskRun → Unknown/Pending
            "pod status \"Ready\":\"False\"; message: \"containers with unready status: [step-use-trusted-artifact step-build step-push
  step-sbom-syft-generate step-prepare-sboms step-upload-sbom]\""

  10:05:30  TaskRun → True/Succeeded
            "All Steps have completed executing"

```
  PipelineRun controller logs:
  ```
10:01:19  PipelineRun status → Running
            "Tasks Completed: 3 (Failed: 0, Cancelled 0), Incomplete: 15, Skipped: 0"

  10:01:20  PipelineRun status → Failed   ← reacted to the transient TaskRun failure
            "Tasks Completed: 4 (Failed: 1, Cancelled 0), Skipped: 14"

  10:01:20  Event Warning: 'Failed' Tasks Completed: 4 (Failed: 1, Cancelled 0), Skipped: 14
```

Full TaskRun reconciliation logs:
  ```
2026-01-16T10:01:07.436Z  Successfully reconciled taskrun swift-operator-1931c00b959f32d146472770cefe485f-build-container/openstack-tenant with
  status: &apis.Condition{Type:"Succeeded", Status:"Unknown", Reason:"ExceededNodeResources", Message:"TaskRun Pod exceeded available resources"}
  2026-01-16T10:01:18.683Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"Pending"
  2026-01-16T10:01:18.743Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"PodReadyToStartContainers\":\"False\";
  message: \"\""
  2026-01-16T10:01:18.800Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"PodReadyToStartContainers\":\"False\";
  message: \"\""
  2026-01-16T10:01:18.812Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"PodReadyToStartContainers\":\"False\";
  message: \"\""
  2026-01-16T10:01:18.824Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"PodReadyToStartContainers\":\"False\";
  message: \"\""
  2026-01-16T10:01:18.918Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"PodReadyToStartContainers\":\"False\";
  message: \"\""
  2026-01-16T10:01:19.925Z  Successfully reconciled taskrun ... Status:"False", Reason:"Failed", Message:"init container failed, \"prepare\" exited
  with code 137: ContainerStatusUnknown"
  2026-01-16T10:01:20.576Z  taskrun openstack-tenant/swift-operator-1931c00b959f32d146472770cefe485f-build-container is still running
  2026-01-16T10:01:20.667Z  Successfully reconciled taskrun ... Status:"Unknown", Reason:"ExceededNodeResources", Message:"TaskRun Pod exceeded
  available resources"
  2026-01-16T10:01:25.367Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"Pending"
  2026-01-16T10:01:25.513Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"PodReadyToStartContainers\":\"False\";
  message: \"\""
  2026-01-16T10:01:32.557Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"Initialized\":\"False\"; message:
  \"containers with incomplete status: [place-scripts]\""
  2026-01-16T10:01:33.475Z  Successfully reconciled taskrun ... Reason:"Pending", Message:"pod status \"Ready\":\"False\"; message: \"containers
  with unready status: [step-use-trusted-artifact step-build step-push step-sbom-syft-generate step-prepare-sboms step-upload-sbom]\""
  2026-01-16T10:05:30.985Z  Successfully reconciled taskrun ... Status:"True", Reason:"Succeeded", Message:"All Steps have completed executing"
```

PipelineRun controller logs:
  ```
2026-01-16T10:01:19.018Z  PipelineRun swift-operator-openstack-18-0-on-pull-request-nvvsc status is being set to &{Succeeded Unknown {2026-01-16
  09:55:34 +0000 UTC} Running Tasks Completed: 3 (Failed: 0, Cancelled 0), Incomplete: 15, Skipped: 0}
  2026-01-16T10:01:19.578Z  PipelineRun swift-operator-openstack-18-0-on-pull-request-nvvsc status is being set to &{Succeeded Unknown {2026-01-16
  09:55:34 +0000 UTC} Running Tasks Completed: 3 (Failed: 0, Cancelled 0), Incomplete: 15, Skipped: 0}
  2026-01-16T10:01:20.620Z  Event(PipelineRun "swift-operator-openstack-18-0-on-pull-request-nvvsc"): type: 'Warning' reason: 'Failed' Tasks
  Completed: 4 (Failed: 1, Cancelled 0), Skipped: 14
  2026-01-16T10:01:20.620Z  PipelineRun swift-operator-openstack-18-0-on-pull-request-nvvsc status is being set to &{Succeeded False {2026-01-16
  10:01:20.619475111 +0000 UTC} Failed Tasks Completed: 4 (Failed: 1, Cancelled 0), Skipped: 14}
```

  Impact

  - PipelineRun remains permanently Failed even though the TaskRun succeeded
  - 14 dependent tasks were skipped unnecessarily
  - Controllers (results, chains, kubearchive) started reconciling the false failure

### PR DESCRIPTION
**Pull request by ab-ghosh**
# Changes
Fixes #9467 - Verify TaskRun failure against the API server in `getTaskRunFunc`, the earliest point where TaskRun data enters the pipeline state. All downstream logic automatically sees the verified data.

When a pod gets evicted, a race condition causes the PipelineRun to permanently fail even though the TaskRun recovers:
  1. Pod eviction → TaskRun controller sets status to `Failed`
  2. Kubernetes reschedules pod → TaskRun controller updates status back to `Running`
  3. **Race**: Between steps 1 and 2, the PipelineRun controller reads the stale `Failed` status from the informer cache and irreversibly marks itself as `Failed`

**Before**
  Informer cache returns TaskRun with stale Failed status (pod was evicted but is being rescheduled)
    → Stale Failed TaskRun is stored in ResolvedPipelineTask.TaskRuns
    → isFailure() returns true → IsStopping() returns true → no new TaskRuns created
    → GetPipelineConditionStatus() returns Failed
    → PipelineRun is permanently marked Failed
    → IsDone() short-circuits all future reconciles — even after TaskRun recovers

**After**
   Informer cache returns TaskRun with stale Failed status
    → getTaskRunFunc detects Failed condition
    → 1 additional API server GET fetches the actual TaskRun status
    → API server returns the real status (Pending/Running/Succeeded)
    → Fresh TaskRun from API server is used as-is.
    → All downstream logic (isFailure(), IsStopping(), GetPipelineConditionStatus()) sees the real status
    → PipelineRun continues normally

Cost: One extra API server GET per failed TaskRun per reconcile. Zero overhead when TaskRuns are successful or running.

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -402,7 +402,27 @@ func (c *Reconciler) resolvePipelineState(
...
+			tr, err := c.taskRunLister.TaskRuns(pr.Namespace).Get(name)
+			if err != nil {
+				return nil, err
+			}
+			// If the informer cache shows the TaskRun as Failed, verify against the API server
+			// to guard against stale cache data causing premature PipelineRun failure.
+			// This addresses a race condition where a TaskRun transiently fails (e.g., pod eviction)
+			// and recovers, but the PipelineRun controller reads the stale Failed status.
+			if cond := tr.Status.GetCondition(apis.ConditionSucceeded); cond != nil && cond.IsFalse() {
+				freshTR, freshErr := c.PipelineClientSet.TektonV1().TaskRuns(pr.Namespace).Get(ctx, name, metav1.GetOptions{})
```
Comment (khrm) [inline pkg/reconciler/pipelinerun/pipelinerun.go:439]: Isn't this too costly for just fixing a bug? I think we need to benchmark this. Our earlier optimisations were used to address this.

**Review by khrm — COMMENTED** [review]
Isn't it possible to just somehow cover the evicted case in the TaskRun reconciler?

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 1 above)
Comment (ab-ghosh) [reply pkg/reconciler/pipelinerun/pipelinerun.go:439]: yes, the cost here is one API server GET call per failed TaskRun during each reconcile loop.

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -402,7 +402,27 @@ func (c *Reconciler) resolvePipelineState(
...
+				if freshErr != nil {
+					logging.FromContext(ctx).Warnf("Cannot verify TaskRun %s status from API server: %v; using cached status", name, freshErr)
+					return tr, nil
```
Comment (vdemeester) [inline pkg/reconciler/pipelinerun/pipelinerun.go]: When the API server GET fails (transient network error, etc.), falling back to the cached Failed TaskRun causes exactly the premature-failure scenario this PR aims to prevent. Returning an error here would let the reconciler retry on the next cycle instead of acting on potentially stale data:

```suggestion
				if freshErr != nil {
					return nil, fmt.Errorf("cannot verify TaskRun %s failure status against API server: %w", name, freshErr)
				}
```

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 1 above)
Comment (vdemeester) [inline pkg/reconciler/pipelinerun/pipelinerun.go]: Since the informer cache is never updated after the fresh GET, every reconcile cycle will re-issue the API server GET for each failed-in-cache TaskRun until the informer naturally catches up. For a PipelineRun with many tasks under an eviction storm, this could create significant API server load. Worth considering either a requeue-with-delay or at minimum a comment acknowledging the trade-off.

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 3]
```diff
@@ -402,7 +402,27 @@ func (c *Reconciler) resolvePipelineState(
...
+				}
+				freshCond := freshTR.Status.GetCondition(apis.ConditionSucceeded)
+				if freshCond == nil || !freshCond.IsFalse() {
```
Comment (vdemeester) [inline pkg/reconciler/pipelinerun/pipelinerun.go:454]: Minor: `freshCond == nil` (unset/unknown status) being treated as "not failed" is defensively reasonable, but an inline comment explaining the reasoning would help future readers.

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 4]
```diff
@@ -402,7 +402,27 @@ func (c *Reconciler) resolvePipelineState(
...
+					logging.FromContext(ctx).Infof("TaskRun %s appears failed in cache but is not failed in API server; using API server status", name)
+					return freshTR, nil
+				}
+			}
+			return tr, nil
 		}
 
 		getCustomRunFunc := func(name string) (*v1beta1.CustomRun, error) {
```
Comment (vdemeester) [inline pkg/reconciler/pipelinerun/pipelinerun.go:464]: Could CustomRuns suffer the same stale-cache problem after eviction? If they're intentionally excluded (e.g., CustomRuns don't create pods, so eviction doesn't apply), a brief comment here explaining why would be helpful.

**Review by vdemeester — CHANGES_REQUESTED** [review]
The problem this addresses is real — stale informer cache after pod eviction causing premature PipelineRun failure. The approach of verifying against the API server is reasonable, but the error handling path and operational characteristics need some attention.

**Comment by ab-ghosh** [comment]
@vdemeester Thank you for the suggestions. Addressed all the pointers, could you please take another look and let me know if anything else needs to be updated?

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 5]
```diff
@@ -405,9 +405,42 @@ func (c *Reconciler) resolvePipelineState(
...
+				return nil, err
+			}
+			// If the informer cache shows the TaskRun as Failed, verify against the API server
+			// to guard against stale cache data causing premature PipelineRun failure.
+			// This addresses a race condition where a TaskRun transiently fails (e.g., pod eviction)
+			// and recovers, but the PipelineRun controller reads the stale Failed status.
+			if cond := tr.Status.GetCondition(apis.ConditionSucceeded); cond != nil && cond.IsFalse() {
+				// Skip API server verification for deterministic terminal states that cannot
+				// recover from pod eviction (e.g., user cancellation or timeout).
+				if cond.Reason == v1.TaskRunReasonCancelled.String() || cond.Reason == v1.TaskRunReasonTimedOut.String() {
```
Comment (waveywaves) [inline pkg/reconciler/pipelinerun/pipelinerun.go:419]: The skip-list only covers `Cancelled` and `TimedOut`, but 14+ other deterministic failure reasons are equally irrecoverable: `TaskRunImagePullFailed`, `TaskRunValidationFailed`, `TaskRunResolutionFailed`, `ResourceVerificationFailed`, `PodCreationFailed`, etc. None of these can recover from pod eviction. Consider inverting the list: only verify for reasons that could be eviction-related (`Failed`, `PodEvicted`, or the new failure reasons from #9368).

**Review by waveywaves — COMMENTED** [review]
Core fix is correct. Verifying against the API server at the `getTaskRunFunc` boundary is the right interception point. All four of vdemeester's concerns have been addressed. The remaining concern is unnecessary API GETs for non-eviction failure reasons, which could be significant for matrixed pipelines.

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 5 above)
Comment (ab-ghosh) [reply pkg/reconciler/pipelinerun/pipelinerun.go]: ahh thats right, let me invert the skip-list to an allow-list, so that it will only make calls to the API server for Failed and PodEvicted.

**Comment by ab-ghosh** [comment]
🤖 **AI-Assisted Code Review**
  I used an AI assistant to help review this PR. Here are the findings:

  **Summary**
  This PR fixes a critical race condition where a PipelineRun is permanently marked as Failed if a TaskRun is temporarily marked as Failed due to a PodEvicted event,
  even if the pod is successfully rescheduled and recovers. The fix introduces a surgical, live API query to verify the TaskRun status when the informer cache
  reports a PodEvicted failure, preventing the PipelineRun from prematurely aborting. The implementation is well-scoped, minimizing API server overhead by explicitly
  limiting the live query to this specific, recoverable failure reason.

  **Findings**

  **Positive (things done well)**
   - The performance cost is carefully mitigated. By limiting the live API call strictly to TaskRuns failing with the PodEvicted reason, the "happy path" and
     non-recoverable failures (like script failures or timeouts) do not incur any additional API overhead.
   - Excellent and thorough unit test coverage. The tests successfully account for API errors, NotFound responses, and ensure that other failure types (like
     Cancelled or generic Failed) do not trigger false positive API calls.
   - Great inline comments inside getTaskRunFunc explaining exactly why this specific API call is being made (defending against stale cache) and why it only targets
     pod eviction.

  **Pattern Conformance**
  The code completely adheres to the repository's existing patterns. It uses standard Tekton clients, k8s.io/apimachinery patterns (like apierrors.IsNotFound()), and
  the established logging.FromContext(ctx) pattern used throughout the controller.

  **Race Condition Analysis**
  The changes are concurrency-safe and actually resolve a distributed systems race condition (stale cache vs. actual API server state). The reconciler inherently
  processes events serially per resource key, and the synchronous API query to fetch fresh state safely accesses immutable context without manipulating shared
  mutable state inappropriately.

  Manual Testing Required: No

  Why: The PR includes extensive and deterministic unit test coverage that successfully simulates the informer cache vs. API server disparity, covering all branch logic without needing complex manual cluster eviction simulations.

  **Verdict**
  Approve
  The fix is surgically targeted, handles edge cases gracefully, includes robust tests, and introduces no measurable performance regressions.

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 6]
```diff
@@ -425,9 +425,39 @@ func (c *Reconciler) resolvePipelineState(
...
+			// premature PipelineRun failure. Pod eviction is the only transient failure
+			// that can recover (pod gets rescheduled); all other failure reasons
+			// (Failed, Cancelled, TimedOut, etc.) are deterministic and cannot recover.
+			if cond := tr.Status.GetCondition(apis.ConditionSucceeded); cond != nil && cond.IsFalse() &&
+				cond.Reason == v1.TaskRunReasonPodEvicted.String() {
+				freshTR, freshErr := c.PipelineClientSet.TektonV1().TaskRuns(pr.Namespace).Get(ctx, name, metav1.GetOptions{})
+				if freshErr != nil {
+					// If the TaskRun is not found on the API server, it was deleted;
+					// fall back to the cached status since retrying won't help.
+					if apierrors.IsNotFound(freshErr) {
```
Comment (waveywaves) [inline pkg/reconciler/pipelinerun/pipelinerun.go:446]: Not convinced og the `NotFound` fallback here. In the normal resolution path, a missing `TaskRun` is treated as an absent child object rather than as a stale failed one. This branch changes that only for the `PodEvicted` verification path by forcing the cached failed `tr` to win even when the authoritative API view is 404. Can you explain why a deleted or missing `TaskRun` should preserve the stale failure instead of falling back to the usual missing-child semantics?

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 7]
```diff
@@ -425,9 +425,39 @@ func (c *Reconciler) resolvePipelineState(
...
+					return nil, fmt.Errorf("cannot verify TaskRun %s failure status against API server: %w", name, freshErr)
+				}
+				freshCond := freshTR.Status.GetCondition(apis.ConditionSucceeded)
+				// Treat nil condition (status not yet set) as non-terminal;
+				// the TaskRun may still be initializing after being rescheduled.
+				if freshCond == nil || !freshCond.IsFalse() {
+					logging.FromContext(ctx).Infof("TaskRun %s appears failed in cache but is not failed in API server; using API server status", name)
+					return freshTR, nil
+				}
+			}
```
Comment (waveywaves) [inline pkg/reconciler/pipelinerun/pipelinerun.go:458]: After paying for the API GET, should we always use `freshTR` on success? Right now we only switch to it when the fresh condition is non-failed; if the fresh object is still `Succeeded=False`, we fall back to the cached `tr`. That means downstream status and reason consumers can still observe stale condition data even though we already fetched the authoritative object.

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 6 above)
Comment (ab-ghosh) [reply pkg/reconciler/pipelinerun/pipelinerun.go:446]: The inline comment is a bit weak, i will update that, The reasoning behind the behavior this is when pod gets evicted, recovery path never deletes the TaskRun's object, controller reschedules a new pod and updates the same TaskRun in-place, So if the API server returns NotFound for a TaskRun that the cache shows as PodEvicted, it can't be a recovery scenario which we are addressing in this pr. In that case, using the cached PodEvicted failure as the final state might be right, let me know if i am missing anything here.

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 7 above)
Comment (ab-ghosh) [reply pkg/reconciler/pipelinerun/pipelinerun.go:458]: yeah make sense, will fix this, we should always return freshTR after a successful GET

**Review by waveywaves — APPROVED** [review]
Thanks, this looks ready to me now. The live API verification is limited to the recoverable `PodEvicted` path, successful GETs now return the fresh TaskRun, NotFound/error behavior is documented and covered, and non-eviction failures avoid the extra API call.

I ran the targeted tests locally:

```sh
go test ./pkg/reconciler/pipelinerun -run "TestReconcile_(DeferFailureWhenTaskRunRecoveredInAPIServer|ConfirmFailureWhenTaskRunFailedInAPIServer|APIServerErrorReturnsError|TaskRunNotFoundFallsBackToCache|CancelledTaskRunNoFalsePositive|TimedOutTaskRunNoFalsePositive|GenericFailedTaskRunSkipsAPIVerification)" -count=1
```

/approve
/lgtm

**Comment by waveywaves** [comment]
Follow-up tracking issue opened for limiting the PodEvicted live TaskRun GET to cases where the TaskRun can still retry: #10102

**Comment by mathur07** [comment]
@ab-ghosh, you might want to cherry-pick this to 1.12
